flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/OperatorCoordinatorHolder.java

```diff
@@ -312,25 +423,97 @@ private void checkpointCoordinatorInternal(
+                                            checkpointId, coordinatorCheckpoint);
+                                } catch (Throwable t) {
+                                    ExceptionUtils.rethrowIfFatalErrorOrOOM(t);
+                                    result.completeExceptionally(t);
+                                    globalFailureHandler.handleGlobalFailure(t);
+                                }
+                            }
+                            return null;
+                        },
+                        mainThreadExecutor);
+    }
+
+    private void abortAllPendingAcknowledgeCloseGatewayFutures(String message) {
+        if (acknowledgeCloseGatewayFutureMap.isEmpty()) {
+            return;
+        }
+
+        for (int subtask : acknowledgeCloseGatewayFutureMap.keySet()) {
+            completeAcknowledgeCloseGatewayFutureExceptionally(subtask, message, null);
+        }
+    }
+
+    private void completeAcknowledgeCloseGatewayFutureExceptionally(
+            int subtask, String message, @Nullable Throwable reason) {
+        if (acknowledgeCloseGatewayFutureMap.containsKey(subtask)) {
+            Exception exception = new FlinkException(message, reason);
+            acknowledgeCloseGatewayFutureMap.remove(subtask).completeExceptionally(exception);
+        }
+    }
+
+    private void completeAcknowledgeCloseGatewayFuture(int subtask, long checkpointId) {
+        // The coordinator holder may receive an acknowledgement event after the checkpoint
+        // corresponding to the event has been aborted, or even after a new checkpoint has started.
+        // The acknowledgement event should be ignored in these cases.
+        if (!isCurrentPendingCheckpoint(checkpointId)) {
+            return;
+        }
+
+        if (acknowledgeCloseGatewayFutureMap.containsKey(subtask)) {
+            acknowledgeCloseGatewayFutureMap.remove(subtask).complete(Acknowledge.get());
+        }
+    }
+
+    /**
+     * Checks whether a provided checkpoint id corresponds to the current pending checkpoint.
+     *
+     * @return true if the provided id corresponds to the current checkpoint, false if the id
+     *     corresponds to a previous checkpoint, or there is no pending checkpoint currently.
+     * @throws IllegalArgumentException if the coordinator holder has never a checkpoint with the
+     *     provided id.
+     */
+    private boolean isCurrentPendingCheckpoint(long checkpointId) {
+        if (checkpointId > latestAttemptedCheckpointId) {
+            throw new IllegalArgumentException(
+                    "The provided checkpoint id "
+                            + checkpointId
+                            + " is related to a newer checkpoint that is unknown to the coordinator holder.");
+        }
+
+        if (currentPendingCheckpointId == NO_CHECKPOINT) {
+            return false;
+        } else {
+            if (latestAttemptedCheckpointId != currentPendingCheckpointId) {
```

Comment:
Neither `latestAttemptedCheckpointId` nor `currentPendingCheckpointId` is updated in this method. Do we need to make this check here?